### ISSUE TITLE
Major Helm plugin security issue

### ISSUE THREAD

**Issue by braderhart** [helm/helm#10316]
Output of `helm version`: 3.7.1
Output of `kubectl version`: 1.22.3

There is a major problem with how plugins are installed and updated. It is basically the same as piping a foreign bash script into your shell. An exploited user may not even know, but they could easily have their entire production Kubernetes clusters exposed, passwords, and more.

For example, take the `helm-push` plugin:

`plugin.yaml`
```
name: "cm-push"
version: "0.10.1"
usage: "Please see https://github.com/chartmuseum/helm-push for usage"
description: "Push chart package to ChartMuseum"
command: "$HELM_PLUGIN_DIR/bin/helm-cm-push"
downloaders:
- command: "bin/helm-cm-push"
  protocols:
  - "cm"
useTunnel: false
hooks:
  install: "cd $HELM_PLUGIN_DIR; scripts/install_plugin.sh"
  update: "cd $HELM_PLUGIN_DIR; scripts/install_plugin.sh"
```

As you can see, it is blanket running a script during install `scripts/install_plugin.sh` as the user. Let's say a malicious user wanted to get your `~/.kube` credentials and `~/.config/glcoud`. All they have to do is include commands to silently upload those. This exposes all files that user has access to, can add commands to run on user login and more. A malicious actor could even have a useful plugin, but with an undisclosed exploit.

Users installing plugins are not going to be aware that simply installing a plugin could do something as damaging as remove their entire user directory, and it shouldn't be allowed. Kubernetes is a big ecosystem, and I doubt it would be that hard to convince users to install your Helm plugin. A user running `helm plugin install https://github.com/helmawesome/helm-awesome-plugin.git` will have no clue that by doing so, they could have just exposed their entire system.

**Comment by bacongobbler** [comment]
Hi @braderhart.

When you install a third-party plugin, it is your responsibility to audit the code and ensure it is not malicious. We've discussed this with several independent security audit teams and they all agreed that it is the responsibility of the user to audit any third party code before installing it. In other words this is a known risk of using "unofficial" code and is therefore a non-issue. If Helm was distributing this code then it'd be a different story, but this is effectively the same as adding a third party repository to apt and calling `apt install totally-not-a-bitcoin-miner`.

in other words, Helm is not responsible for auditing third party code. That is on you.

In the future, please make sure to disclose security issues responsibly. See both https://github.com/helm/helm/blob/main/CONTRIBUTING.md#reporting-a-security-issue as well as https://github.com/helm/helm/blob/main/SECURITY.md for more information. Thanks!

**Comment by bacongobbler** [comment]
Closing as a non-issue.

**Comment by braderhart** [comment]
@bacongobbler The issue with privately reporting security vulnerabilities is that there is no public record. Because this appeared to be more of a willful choice, and as you've stated this is a non-issue, what precisely would be the benefit of being told something is a non-issue privately? At least this way, if it does become an issue it will be publicly documented that someone brought it to the attention of the Helm authors before and they declared it a non-issue.

Here is an example of a plugin install:

```
helm plugin install https://github.com/chartmuseum/helm-push.git
Downloading and installing helm-push v0.10.1 ...
https://github.com/chartmuseum/helm-push/releases/download/v0.10.1/helm-push_0.10.1_linux_amd64.tar.gz
Installed plugin: cm-push
```

There is no prompt before installing. `helm plugin install --help` shows no message about auditing third party code. While you may have talked to third-party auditors about this before, there is no prompted message to the user; nor any indication that installing a plugin allows a plugin author to run scripts as that user. There are plenty of actions Helm could take to restrict the attack vector here, but that isn't being done. I wouldn't have even reported the issue if there was a prompt at least informing the user.

I can't predict what would happen, but let's say hypothetically this does become an issue and a major enterprise does end up getting attacked due to this problem. A bunch of articles come out describing how they were exploited. What do you think the general consensus is going to be in the security community? That users should have just known installing a Helm plugin allowed them to be exploited? No warning message, but they should have predicted it?

I find your reasoning here concerning because you act like there is nothing that can be done here to improve the security, even if it is prompting the user on install.

**Comment by bacongobbler** [comment]
> what precisely would be the benefit of being told something is a non-issue privately? At least this way, if it does become an issue it will be publicly documented that someone brought it to the attention of the Helm authors before and they declared it a non-issue.

The goal of responsibly disclosing security issues is to provide maintainers time to write a fix, test it, announce a new security release to the community on the mailing list, issue a CVE, and cut the release. It also gives maintainers a bit of time to assess the situation and test fixes over regular business hours.

Consider the situation: had this actually been a major [zero-day](https://en.wikipedia.org/wiki/Zero-day_(computing), by disclosing it publicly you now have given attackers an opportunity to write exploits well before the maintainers have an opportunity to write and release a fix. And the maintainers would've been called in on their Friday night to work on a fix well into the weekend.

> Here is an example of a plugin install:

Again, you are just arguing the semantics of what happens when you call `helm install https://github.com/bacongobbler/totally-not-a-bitcoin-miner`. Even if there was no `exec-hook` available, I could just as easily embed the bitcoin miner [in the plugin's source code](https://therecord.media/crypto-miner-found-hidden-inside-three-npm-libraries/). If you have ideas or suggestions on how to improve this, please feel free to write a HIP as described in https://github.com/helm/helm/blob/main/CONTRIBUTING.md#proposing-an-idea.

If there was a way to either

a) coerce the user to install a plugin without them explicitly calling `helm plugin install`, or
b) exploit a `helm plugin install` to execute an RCE,

Then it's a security issue in Helm. Otherwise it's third-party (read: untrusted) code.

> I can't predict what would happen

That's on you. You're responsible for auditing any code you use.

**Comment by braderhart** [comment]
@bacongobbler 

> Consider the situation: had this actually been a major zero-day, by disclosing it publicly you now have given attackers an opportunity to write exploits well before the maintainers have an opportunity to write and release a fix. And the maintainers would've been called in on their Friday night to work on a fix well into the weekend.

If this had been a 0-day with Helm itself, then yes I would have used responsible disclosure. Instead, this is a willful decision that makes exploiting users relatively easy.

> Again, you are just arguing the semantics of what happens when you call helm plugin install https://github.com/bacongobbler/totally-not-a-bitcoin-miner. Even if there was no exec-hook available, I could just as easily embed the bitcoin miner in the plugin's source code.

There is a clear distinction between what happens when you install something and run something. I don't disagree that that there a lot of ways a plugin author could run an exploit once you install a plugin, and there could be improvements to address those—but that would require a lot of extra work and is not practical without a new major release and deprecation notice.

I may consider writing a HIP. However, I only have so much time to point out obvious issues mixed with family, work, etc. I understand that is part of open source, so I'm willing to do if it will get done, but if I am going to spend time writing something up to have it get dismissed as "intended" behavior then I don't see the point.

**Comment by bacongobbler** [comment]
> There is a clear distinction between what happens when you install something and run something.

Every package manager has a mechanism of a pre-install hook. dpkg has both `dpkg::Pre-Invoke` and `dpkg::Post-Invoke`, both of which are called without user interaction after calling `dpkg -i` or `apt install`. Helm plugin install hooks are no different.
